internal/dag/dag.go

```diff
@@ -18,13 +18,17 @@ limitations under the License.
 package dag
 
 import (
+	"github.com/Masterminds/semver"
 	"github.com/crossplane/crossplane-runtime/pkg/errors"
 )
 
 // Node is a node in DAG.
 type Node interface {
 	Identifier() string
 	Neighbors() []Node
+	GetConstraints() string
+	GetParentConstraints() []string
```

Comment:
💯 to what @turkenh said. I spent the evening refreshing my memory on how package dependency resolution works, and I kept having to go find the implementation of this interface to figure out what its methods actually did. It would be a big help to (retroactively) document it, if possible.

I also wonder whether it'd be more readable if the DAG and Node interface methods were more specific to the use case - i.e. dependencies. For example if `TraceNode` is really returning all of a node's dependencies it might be more readable to call it `Dependencies`. That's way out of scope for this PR though. 🙂

Reply:
> That's way out of scope for this PR though

Actually, I'm now questioning whether it should be in scope for this PR. Especially in the context of the DAG now being even more coupled to the only use case we have for it - i.e. dependency solving. It's already pretty leaky now that it has e.g. `GetConstraints` as a method.